add-lab_1

MChepulis opened the pull request:
add build script
add lab_1

MChepulis committed:
add-lab_1

add build script
add lab_1

KotsMikhail reviewed (changes requested):
Если указать два одинаковых пути в конфиге, то демон удалит из этой папки все файлы
Если запустить демона, не указав в конфиге интервал, то, судя по всему, демон запускается с интервалом 0.
Неодинаковый стиль отступов фигурных скобок
Как так получилось, что ваша работа процентов на 50 совпадает с работой Дмитрия Федорова, включая сообщения об ошибках?

MChepulis committed:
Update lab_1.cpp

add same path check
add frequency positive (>=0) check
fix braces spaces

MChepulis commented:
исправил отступы фигурных скобок
добавил ограничения на считываемые из config файла данные:
  частота должна быть больше нуля
  пути директорий должны различаться

KotsMikhail commented on chepulis.mikhail/lab_1/lab_1.cpp:
```diff
@@ -0,0 +1,303 @@
+
+#define PID_FILENAME "pid_info.txt"
+#define PROC_DIR "/proc"
+#define WAIT_OF_KILLING_TIME 1
```
Не стоит использовать директивы препроцессора для объявления констант

KotsMikhail commented on chepulis.mikhail/lab_1/lab_1.cpp:
```diff
@@ -0,0 +1,303 @@
+  }
+  rv = remove(fpath);
+  if (rv) {
+    perror(fpath);
```
Что делает эта функция и зачем она тут?

MChepulis replied:
perror(fpath) - выводит сообщение о последней ошибке произошедшей в функции remove с аргументом fpath, где аргумент fpath будет выведен до текста ошибки (для получения большей информации)
является отладочной информацией
по сути не требуется также как и все закомментированные std::cout, т.к. perror посылает сообщение в stderr, который в демоне отключён

KotsMikhail commented on chepulis.mikhail/lab_1/lab_1.cpp:
```diff
@@ -0,0 +1,303 @@
+  rv = remove(fpath);
+  if (rv) {
+    perror(fpath);
+    syslog(LOG_USER, "Something goes wrong while deleating files \"%s\"", fpath);
```
Почему ошибки в работе помечаются флагом USER?

KotsMikhail commented on chepulis.mikhail/lab_1/lab_1.cpp:
```diff
@@ -0,0 +1,303 @@
+      //std::cout << "Terminate daemon by signal \n";
+      unlink(pid_file_path.c_str());
+      closelog();
+      exit(SIGTERM);
```
Что код сигнала делает в аргументе функции exit?

KotsMikhail commented on chepulis.mikhail/lab_1/lab_1.cpp:
```diff
@@ -0,0 +1,303 @@
+  syslog(LOG_USER, "Deamon already works - kill him");
+  //std::cout << "Deamon already works - kill him\n";
+  if ((dir = opendir(pidFolder.c_str())) != NULL) {
+    kill(atoi(pid.c_str()), SIGTERM);  
```
Что будет если послать сигнал несуществующему процессу?

MChepulis replied:
ничего, но если послать с pid="0" то убьём сами себя (т.к. Если pid равен 0, то sig отправляется каждому процессу в группе процессов вызывающего процесса)
можно ввести проверку существования данного pid путем послания по нему сигнал 0 и проверки возвращаемого значения

KotsMikhail commented on chepulis.mikhail/lab_1/lab_1.cpp:
```diff
@@ -0,0 +1,303 @@
+}
+
+
+bool CompareExtantion(const std::string fileName, const std::string extantion){
```
Слово "расширение" пишется не так

KotsMikhail commented on chepulis.mikhail/lab_1/lab_1.cpp:
```diff
@@ -0,0 +1,303 @@
+  DIR*           dir;
+  struct dirent* ent;
+  std::vector<std::string> res;
+  if ((dir = opendir(absPath.c_str())) != NULL) {
```
Лучше использовать nullptr вместо NULL

KotsMikhail commented on chepulis.mikhail/lab_1/lab_1.cpp:
```diff
@@ -0,0 +1,303 @@
+    }
+    closedir(dir);
+  } else {
+  return std::vector<std::string>();
```
С отступами всё ещё проблемы в некоторых местах

KotsMikhail commented on chepulis.mikhail/lab_1/lab_1.cpp:
```diff
@@ -0,0 +1,303 @@
+    return -1;
+  }
+
+  int pid = fork();
```
Почему 1-ое ветвление в main, а все остальное - в функции демонизации?

KotsMikhail commented on chepulis.mikhail/lab_1/lab_1.cpp:
```diff
@@ -0,0 +1,303 @@
+void signalHandler(int signal){
+  switch (signal) {
+    case SIGHUP:{ 
+      is_need_work = false;
```
Что может пойти не так, если не останавливать в этом месте работу демона?

MChepulis replied:
запрещает запуск работы демона, пока считывается файл
если не запрещать, то демон может начать работы пока файл до конца не считан и произойдёт некорректное поведение (например уже новая директория-источник, но ещё старая директория-назначения)

KotsMikhail replied:
Допустим, вам прилетел сигнал sighup между чтением содержимого директории и проверкой расширения (например, где-нибудь в строке 271 этого файла). Как вы считаете, остановит ли такая защита выполнение вашей программы?
А если сигнал придет точно перед проверкой в строке 288, гарантирует ли это, что программа не зайдет в тело цикла пока не закончит обработку сигнала?

MChepulis replied:
Флаг is_need_work лишь запрещает запускать новый цикл обработки, никак не влияя на тот, что уже идёт

т.к. сейчас нет никаких проверок идёт ли обработка прямо сейчас, то по сигналу SIGHUP перезапишутся переменные src_dir и dst_dir
в лучшем случае мы получим ошибку копирования ( и просто пропустим оставшиеся файлы), т.к. мы будем искать файлы из "старой" директории в "новой"
в худшем - мы продолжим копирование оставшиеся (те что ещё не скопировали) файлы в новую директорию-назначения, притом никто не гарантирует, что новая директория пустая и в ней есть нужные папки (IMG, OTHER)

KotsMikhail commented on chepulis.mikhail/lab_1/lab_1.cpp:
```diff
@@ -0,0 +1,303 @@
+      break;
+    }
+    case SIGTERM:{
+      is_need_work = false;
```
Что может пойти не так, если не останавливать в этом месте работу демона?

MChepulis replied:
запрещает запуск работы демона, пока считывается файл
если не запрещать, то демон может начать работы пока файл до конца не считан и произойдёт некорректное поведение (например уже новая директория-источник, но ещё старая директория-назначения)

MChepulis replied:
тут запрет запуска работы демона ни на что не влияет
нужен лишь для красоты (получил сигнал -> запретил начинать работу -> обработал сигнал) аналогично break после return в теле swith`а

KotsMikhail commented on chepulis.mikhail/lab_1/lab_1.cpp:
```diff
@@ -0,0 +1,303 @@
+      is_need_work = false;
+      syslog(LOG_USER, "Terminate daemon by signal");
+      //std::cout << "Terminate daemon by signal \n";
+      unlink(pid_file_path.c_str());
```
В случае завершения работы демона с ошибкой удаления pid файла не происходит

KotsMikhail commented on chepulis.mikhail/lab_1/lab_1.cpp:
```diff
@@ -0,0 +1,303 @@
+  if (ftwbuf->level == 0) {
+    return 0;
+  }
+  rv = remove(fpath);
```
В программе вы используете 2 разных функции для удаления unlink и remove. В чем между ними разница, и чем обусловлен ваш выбор?

MChepulis replied:
unlink - там где точно знаем, что удаляем файл (чуть быстрее чем remove)
remove - не знаем, удаляется файл или директория(сам вызовет для файлов unlink, для директорий rmdir)

MChepulis committed:
Update lab_1.cpp

Перенес константы из деректив препроцессора в глобальные переменные

справил флаг LOG_USER на флаг LOG_ERR там, где в лог выводится информация об ошибке

исправил SIGTERM на EXIT_SECCESS

перенёс ветвление из  main  в Init

добавил функцию удаления pid-файла и её вызов при завершении работы с ошибкой

заменил NULL на nullptr

исправил отступ

исправил ошибку в названии функции "CompareExtantion"

MChepulis commented:
Перенес константы из деректив препроцессора в глобальные переменные

справил флаг LOG_USER на флаг LOG_ERR там, где в лог выводится информация об ошибке

исправил SIGTERM на EXIT_SECCESS

перенёс ветвление из  main  в Init

добавил функцию удаления pid-файла и её вызов при завершении работы с ошибкой

заменил NULL на nullptr

исправил отступ

исправил ошибку в названии функции "CompareExtantion"

KotsMikhail commented on chepulis.mikhail/lab_1/lab_1.cpp:
```diff
@@ -0,0 +1,309 @@
+int update_freq;
+bool is_need_work;
+
+void delete_pid_file(){
```
Зачем выносить это в отдельную функцию?

MChepulis replied:
вынес для удобства
чтобы легко добавлять отладочную информацию при закрытии файла

KotsMikhail commented on chepulis.mikhail/lab_1/lab_1.cpp:
```diff
@@ -0,0 +1,309 @@
+}
+
+
+bool CompareExtention(const std::string fileName, const std::string extantion){
```
Теперь имя функции правильное, имя переменной - нет

MChepulis committed:
Update lab_1.cpp

rename extantion to extantion

MChepulis committed:
Update lab_1.cpp

проверил в словаре, правильное правописание "exstension"

MChepulis commented:
Исправил ошибку в названии переменной extension

SWAGLORD1 commented:
Вы все ещё не ответили на мои вопросы по поводу переменной need_work

MChepulis commented:
(отвечено ранее)
[
Флаг is_need_work лишь запрещает запускать новый цикл обработки, никак не влияя на тот, что уже идёт

т.к. сейчас нет никаких проверок идёт ли обработка прямо сейчас, то по сигналу SIGHUP перезапишутся переменные src_dir и dst_dir
в лучшем случае мы получим ошибку копирования ( и просто пропустим оставшиеся файлы), т.к. мы будем искать файлы из "старой" директории в "новой"
в худшем - мы продолжим копирование оставшиеся (те что ещё не скопировали) файлы в новую директорию-назначения, притом никто не гарантирует, что новая директория пустая и в ней есть нужные папки (IMG, OTHER)
]

нет, выполнение оно не остановит, т.к. данный флаг влияет только на запуск нового цикла обработки

если сигнал придёт в строке 271, то перезапишутся переменные src_dir и dst_dir
в лучшем случае мы получим ошибку копирования ( и просто пропустим оставшиеся файлы), т.к. мы будем искать файлы из "старой" директории в "новой"
в худшем - мы продолжим копирование оставшиеся (те что ещё не скопировали) файлы в новую директорию-назначения, притом никто не гарантирует, что новая директория пустая и в ней есть нужные папки (IMG, OTHER)

если прямо в строке 288, то тоже не гарантируется то, что запуск не произойдёт, т.к. мы можем проверить значение флага, до того, как оно изменится в signalHadler`е
и тут возможно 2 варианта:
 *  первый - в процессе работы перезаписываются критические переменные src_dir и dst_dir, он будет аналогичен предыдущему случаю (сигнал пришёл в строке 271)
 *  второй же - нам очень повезло, signalHadler успел изменить переменные и мы начинаем работу уже с "новыми" директориями

Данная защита рассчитана на то, что время работы крайне мало по сравнения с временем ожидания.
а значит пользователю проблематично послать сигнал так, чтобы он испортил работу

Данный флаг является лишь примитивной защитой
для полной защиты пришлось бы делать семафор (т.к. переменные src_dir и dst_dir являются критическими ресурсами)
также, чтобы уменьшить зону критического кода, в функции обработки переменные src_dir и dst_dir стоит сделать локальными. тогда критическими будут только 2 строчки - считывание переменные, а не весь цикл обработки

SWAGLORD1 commented:
В целом, вы правы, кроме одного принципиального момента. Обработчики сигналов выполняются синхронно с основной программой. Посылка сигнала генерирует аппаратное прерывание, которое меняет номер текущей исполняемой инструкции. Вся ваша программа исполняется в одном потоке, никакой параллельности тут нет. Единственная проблема, которая в теории может возникнуть - это прерывание обработчика сигнала другим сигналом, но в этой лабораторной это ничего не должно испортить. Так что никакой защиты тут не нужно.

MChepulis commented:
что-то ещё требуется исправить/объяснить для сдачи лабораторной?

SWAGLORD1 commented:
Нужно убрать переменную need_work, так как она не делает ничего полезного

MChepulis committed:
remove is_need_work  flag

remove is_need_work  flag

MChepulis commented:
Убрал переменную-флаг is_need_work

KotsMikhail reviewed (changes requested):
Нет, не убрали, она всё ещё осталась в двух местах

MChepulis committed:
remove is_need_work

удалил остатки переменной is_need_work

MChepulis commented:
удалил объявление переменной is_need_work
удалил is_need_work из функции CreateDaemon

SWAGLORD1 commented:
Засчитано
